### PR TITLE
Contant message processing via single __process() method (improves Notification via HTTP support)

### PR DESCRIPTION
**Pull request by ekzobrain**
Also add tests for Django 1.11

### CODE REVIEW

File: channels_jsonrpc/jsonrpcconsumer.py [hunk 1]
```diff
@@ -325,59 +323,12 @@ def notify_channel(cls, reply_channel, method, params):
...
-    def __process_notification(cls, data, original_msg):
+    def __process(cls, data, original_msg, is_notification=False):
         """
-                Process an inbound JSON-RPC notificaton
-                :param data: object
-                :return: object
-                """
-
-        if 'method' not in data:
-            logger.warning("The notification doesn't have a 'method'")
```
Comment (millerf) [inline channels_jsonrpc/jsonrpcconsumer.py:336]: This is not covered anymore, instead of a warning it sends back a Exception with a null "id".
It needs to be modified

File: channels_jsonrpc/jsonrpcconsumer.py [hunk 2]
```diff
@@ -325,59 +323,12 @@ def notify_channel(cls, reply_channel, method, params):
...
-            return
-
-        method_name = data['method']
-        if not isinstance(method_name, string_types):
-            logger.warning("The notification doesn't have a valid 'method'")
```
Comment (millerf) [inline channels_jsonrpc/jsonrpcconsumer.py:341]: This is not covered anymore, instead of a warning it sends back a Exception with a null "id".
It needs to be modified

File: channels_jsonrpc/jsonrpcconsumer.py [hunk 3]
```diff
@@ -325,59 +323,12 @@ def notify_channel(cls, reply_channel, method, params):
...
-            return
-
-        if method_name.startswith('_'):
-            logger.warning("The method '%s' of the notification cannot be found" % method_name)
```
Comment (millerf) [inline channels_jsonrpc/jsonrpcconsumer.py:345]: This is not covered anymore, instead of a warning it sends back a Exception with a null "id".
It needs to be modified

File: channels_jsonrpc/jsonrpcconsumer.py [hunk 4]
```diff
@@ -325,59 +323,12 @@ def notify_channel(cls, reply_channel, method, params):
...
-            return
-
-        try:
-            method = cls.available_rpc_notifications[id(cls)][method_name]
-            proto = original_msg.channel.name.split('.')[0]
-            if not method.options[proto]:
-                logger.warning("The method '%s' of the notification cannot be found" % method_name)
```
Comment (millerf) [inline channels_jsonrpc/jsonrpcconsumer.py:352]: This is not covered anymore, instead of a warning it sends back a Exception with a null "id".
It needs to be modified

File: channels_jsonrpc/jsonrpcconsumer.py [hunk 5]
```diff
@@ -325,59 +323,12 @@ def notify_channel(cls, reply_channel, method, params):
...
-        except KeyError:
-            # We don't send back anything (notification)
-            pass
-            return
-        params = data.get('params', [])
-
-        if not isinstance(params, (list, dict)):
-            logger.warning("The params '%s' are not valid" % params)
```
Comment (millerf) [inline channels_jsonrpc/jsonrpcconsumer.py:360]: This is not covered anymore, instead of a warning it sends back a Exception with a null "id".
It needs to be modified

**Review by millerf — CHANGES_REQUESTED** [review]

**Comment by ekzobrain** [comment]
All changes you mentioned are intentional. Why should we raise Exceptions for calls, but log warnings for notifications? I think we should raise an Exception both for calls and notifications, and also log warnings for both of them, along with returning result for calls. All calls and notification errors must be logged at server side, regarding of what is sent back to client.
This is what is done now, except logging in case of Exception. But I plan to implement it in a separate task (ovarall improvements to logging).

**Comment by millerf** [comment]
Exception will sends back a JSON-RPC frame, which shouldn’t happen for notifications… And also it looks like (I haven’t tested) the frame sent back has “null” as id…Which is incorrect...

**Comment by ekzobrain** [comment]
No, they won't. Because in methods "http_handler" and "raw_recieve" we already filter out notification responses (which may only occur in case of error, but allow us to take appropriate actions, for example set correct HTTP response code).

**Review by millerf — APPROVED** [review]
